stdlib/LibGit2/src/tree.jl

```diff
@@ -12,27 +12,33 @@ subtree, traversing upwards through it, then traversing the next right subtree
 
 The function parameter `f` should have following signature:
 
-    (Cstring, Ptr{Cvoid}, Ptr{Cvoid}) -> Cint
+    (String, GitTreeEntry) -> Cint
 
 A negative value returned from `f` stops the tree walk. A positive value means
 that the entry will be skipped if `post` is `false`.
 """
-function treewalk(f::Function, tree::GitTree, payload=Any[], post::Bool = false)
-    cbf = cfunction(f, Cint, Tuple{Cstring, Ptr{Cvoid}, Ptr{Cvoid}})
-    cbf_payload = Ref{typeof(payload)}(payload)
+function treewalk(f, tree::GitTree, post::Bool = false)
     # NOTE: don't use @check/GitError directly, because the code can be arbitrary
+    payload = Any[ tree, f ]
+    cbf = @cfunction(function treewalk_callback(root_cstr::Cstring, entry_ptr::Ptr{Cvoid}, payload::Vector{Any})::Cint
```

Comment:
Isn't it possible to capture `tree` and `f` instead of using the `payload` mechanism?

Reply:
Yes, it's just slower that way than to just use the C API correctly

Reply:
I presume it's slower because closure support requires a stub that sets the locals and jumps to the "real" function? Whereas without any closed-over variables, it can just call directly.

Reply:
Yes, and in addition to the indirect function call, it also requires allocation to be able to track the lifetime, and runtime work to manage and handle that, etc.